Add TTL perf caveat on tables with multi indexes

rmloveland opened the pull request:
Especially large tables

Fixes DOC-9465

rmloveland commented:
Hi @annrpom!

I thought you might be the best reviewer based on doing that recent investigation around this

however I understand you may be pretty busy so let me know if I should ask someone else

thanks!

annrpom commented:
Hey - just acking! I'll take a look at this at the EOD today

annrpom commented on src/current/_includes/v23.2/known-limitations/row-level-ttl-limitations.md:
```diff
@@ -1 +1,4 @@
 - Any queries you run against tables with Row-Level TTL enabled do not filter out expired rows from the result set (this includes [`UPDATE`s]({% link {{ page.version.version }}/update.md %}) and [`DELETE`s]({% link {{ page.version.version }}/delete.md %})). This feature may be added in a future release. For now, follow the instructions in [Filter out expired rows from a selection query]({% link { ...
+- Enabling Row-Level TTL on a table with multiple [secondary indexes]({% link {{ page.version.version }}/indexes.md %}) can have negative performance impacts on a cluster, including increased [latency]({% link {{ page.version.version }}/common-issues-to-monitor.md %}#service-latency) and [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contenti ...
+    - The increased latency can occur because deleting data from multiple indexes means multiple [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#architecture-range) must be written to perform the deletion; these ranges may be distributed across different nodes in the cluster. 
```
Deleting from multiple ranges is also true for a table with primary indexes; so, I think it's important to highlight that more indexes = more ranges since secondary indexes don't necessarily lie in the same range and primary indexes.

I'm not sure where in that blurb sounds the best, though

rmloveland replied:
I've rewritten this a bit to try to incorporate your comment and make things clearer (hopefully).

PTAL and let me know what you think!

annrpom commented on src/current/_includes/v23.2/known-limitations/row-level-ttl-limitations.md:
```diff
@@ -1 +1,4 @@
 - Any queries you run against tables with Row-Level TTL enabled do not filter out expired rows from the result set (this includes [`UPDATE`s]({% link {{ page.version.version }}/update.md %}) and [`DELETE`s]({% link {{ page.version.version }}/delete.md %})). This feature may be added in a future release. For now, follow the instructions in [Filter out expired rows from a selection query]({% link { ...
+- Enabling Row-Level TTL on a table with multiple [secondary indexes]({% link {{ page.version.version }}/indexes.md %}) can have negative performance impacts on a cluster, including increased [latency]({% link {{ page.version.version }}/common-issues-to-monitor.md %}#service-latency) and [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contenti ...
```
could we change 

> with terabytes of data and billions of rows

with terabytes of data and billions of rows, split up into multiple ranges & on separate nodes

or something of the sort

rmloveland replied:
Updated to include that info - please take a look!

annrpom reviewed (commented):
Hey - thanks for doing this! I left some comments and I have a trade offer for you: would you be able to review https://github.com/cockroachdb/docs/pull/18208?

Just unsure if I followed procedures correctly :^)

rmloveland commented:
@annrpom this should be RFAL - hopefully latest updates addressed your feedback. Please let me know if not tho!

annrpom commented:
I think this looks good - thank you! Two more things to note:

1. in theory, with the admission control settings mentioned below, we shouldn't see secondary indexes impact latency: 

> New in v23.2: Latency of row-level TTL queries is further reduced by using the elastic CPU limiter, which dynamically controls the total CPU percentage used by row-level TTL reads. The elastic CPU limiter can be disabled for row-level TTL queries by setting the kvadmission.low_pri_read_elastic_control.enabled and sqladmission.low_pri_read_response_elastic_control.enabled [cluster settings](https://deploy-preview-18205--cockroachdb-docs.netlify.app/docs/v23.2/cluster-settings) to false.

I'm not sure if this is redundant or necessary to mention this again here; so, I'll defer to your opinion :^)

2. It would be nice to mention somewhere that these secondary indexes can have a negative impact on the TTL job itself as well (we did some testing Friday that showed that in particular, the  `processing rate` - how many rows per second were being processed - is worse on a table with secondary indexes). We can also suggest here that decreasing the `delete_batch_size` may help by decreasing the amount of ranges that need to be accessed~

rmloveland commented:
> I think this looks good - thank you! 

np! thanks for looking at it

> Two more things to note:
> 
> 1. in theory, with the admission control settings mentioned below, we shouldn't see secondary indexes impact latency:
> 
> > New in v23.2: Latency of row-level TTL queries is further reduced by using the elastic CPU limiter, which dynamically controls the total CPU percentage used by row-level TTL reads. The elastic CPU limiter can be disabled for row-level TTL queries by setting the kvadmission.low_pri_read_elastic_control.enabled and sqladmission.low_pri_read_response_elastic_control.enabled [cluster settings](https://deploy-preview-18205--cockroachdb-docs.netlify.app/docs/v23.2/cluster-settings) to false.
>
> I'm not sure if this is redundant or necessary to mention this again here; so, I'll defer to your opinion :^)

My opinion is that it probably doesn't hurt to say the secondary indexes "may" incur latency.  Especially if it is the case (which, IDK) we haven't yet seen many TTL jobs on tables of this scale on v23.2 yet. 

I've updated the sentence about latency to say "may" just in case. It will be easy to delete the sentence later if it turns out not to be true.

> 2. It would be nice to mention somewhere that these secondary indexes can have a negative impact on the TTL job itself as well (we did some testing Friday that showed that in particular, the  `processing rate` - how many rows per second were being processed - is worse on a table with secondary indexes). We can also suggest here that decreasing the `delete_batch_size` may help by decreasing the amount of ranges that need to be accessed~

I added another bullet to the list with approximately this info (IIUC), please let me know what you think

annrpom reviewed (approved):
awesome - lgtm!

taroface commented on src/current/_includes/v23.2/known-limitations/row-level-ttl-limitations.md:
```diff
@@ -1 +1,5 @@
 - Any queries you run against tables with Row-Level TTL enabled do not filter out expired rows from the result set (this includes [`UPDATE`s]({% link {{ page.version.version }}/update.md %}) and [`DELETE`s]({% link {{ page.version.version }}/delete.md %})). This feature may be added in a future release. For now, follow the instructions in [Filter out expired rows from a selection query]({% link { ...
+- Enabling Row-Level TTL on a table with multiple [secondary indexes]({% link {{ page.version.version }}/indexes.md %}) can have negative performance impacts on a cluster, including increased [latency]({% link {{ page.version.version }}/common-issues-to-monitor.md %}#service-latency) and [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contenti ...
+    - Increased latency may occur because secondary indexes aren't necessarily stored in the same underlying [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#architecture-range) as a table's [primary indexes]({% link {{ page.version.version }}/primary-index.md %}). Further, the secondary indexes' ranges may have [leaseholders]({% link {{ page.version.version }}/architecture ...
```
```suggestion
    - Increased latency may occur because secondary indexes aren't necessarily stored on the same underlying [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#architecture-range) as a table's [primary indexes]({% link {{ page.version.version }}/primary-index.md %}). Further, the secondary indexes' ranges may have [leaseholders]({% link {{ page.version.version }}/architecture/overview.md %}#architecture-leaseholder) located on different nodes than the primary index.
```

rmloveland replied:
Updated!

taroface commented on src/current/_includes/v23.2/known-limitations/row-level-ttl-limitations.md:
```diff
@@ -1 +1,5 @@
 - Any queries you run against tables with Row-Level TTL enabled do not filter out expired rows from the result set (this includes [`UPDATE`s]({% link {{ page.version.version }}/update.md %}) and [`DELETE`s]({% link {{ page.version.version }}/delete.md %})). This feature may be added in a future release. For now, follow the instructions in [Filter out expired rows from a selection query]({% link { ...
+- Enabling Row-Level TTL on a table with multiple [secondary indexes]({% link {{ page.version.version }}/indexes.md %}) can have negative performance impacts on a cluster, including increased [latency]({% link {{ page.version.version }}/common-issues-to-monitor.md %}#service-latency) and [contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contenti ...
+    - Increased latency may occur because secondary indexes aren't necessarily stored in the same underlying [ranges]({% link {{ page.version.version }}/architecture/overview.md %}#architecture-range) as a table's [primary indexes]({% link {{ page.version.version }}/primary-index.md %}). Further, the secondary indexes' ranges may have [leaseholders]({% link {{ page.version.version }}/architecture ...
+    - The contention can occur because [intents]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#write-intents) must be written as part of performing the deletions.
```
```suggestion
    - Increased contention may occur because [intents]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#write-intents) must be written as part of performing the deletions.
```

rmloveland replied:
Much better, thanks - added!

taroface reviewed (approved):
LGTM! Some optional suggestions
